### PR TITLE
Accenture CTI Threat Intelligence Report

### PR DESCRIPTION
**Pull request by satyakidroid**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
We have created this integration to get our reports available in Accenture IntelGraph platform to XSOAR, by creating 
- 2 custom Incident Types
- 2 Layouts
- 12 Incident Fields
We are now able to fetch data and tag them properly in the layout

## Screenshots
image
image
image

## Minimum version of Cortex XSOAR
- [x] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

**Comment by kgal-pan** [comment]
Hey @satyakidroid

- Please remove print statements from Unit Tests as it's causing [failures](https://app.circleci.com/pipelines/github/demisto/content/148881/workflows/6c253c0d-bf1f-4b2e-b0bb-91d40650a5b1/jobs/489661).

- Please run to change latest Docker image of integration:

```bash
demisto-sdk format -ud -i Packs/AccentureCTI/Integrations/ACTIThreatIntelReport/ACTIThreatIntelReport.yml
```

- Please separate the `incidentfields.json` incident fields into separate ones. Same for `layouts.json` and `reputations.json`.

- Please fix inline HTML in `README.md` in line 2. I believe you need to add quotation marks to the `img` `src` attribute with the URL.

**Comment by kgal-pan** [comment]
@satyakidroid - Please:

- Increase test coverage for `ACTIThreatIntelReport`, you can see where the coverage is missed in the code [here](https://490205-60525392-gh.circle-artifacts.com/0/artifacts/coverage_report/html/d_cea6e15193f44573_ACTIThreatIntelReport_py.html).

- Separate the `incidentfields.json` incident fields into separate ones. Same for `layouts.json` and `reputations.json`.

- Seems that the `ip` reputation command is missing `DBotScore.Vendor`, see [here](https://app.circleci.com/pipelines/github/demisto/content/149131/workflows/0384fbb4-fe79-43b3-8a53-83a32052f05e/jobs/490204?invite=true#step-108-82).

**Comment by satyakidroid** [comment]
> @satyakidroid - Please:
> 
> * Increase test coverage for `ACTIThreatIntelReport`, you can see where the coverage is missed in the code [here](https://490205-60525392-gh.circle-artifacts.com/0/artifacts/coverage_report/html/d_cea6e15193f44573_ACTIThreatIntelReport_py.html).
> * Separate the incident fields in `incidentfields.json` into JSON files. Same for `layouts.json` and `reputations.json`.
> * Seems that the `ip` reputation command is missing `DBotScore.Vendor`, see [here](https://app.circleci.com/pipelines/github/demisto/content/149131/workflows/0384fbb4-fe79-43b3-8a53-83a32052f05e/jobs/490204?invite=true#step-108-82).

yep @kgal-pan, looking at these issues now

**Comment by kgal-pan** [comment]
@satyakidroid 

- Why do the `indicatorfields` JSON contain `incidentField` object? 

image

- the incident/indicator field JSON needs to be an object. so for example, using the file above, to get to what the JSON file indicatorfield-ACTI_Threat_Types.json should contain, run:

```bash
❯ jq ".incidentFields[0]" Packs/AccentureCTI/IndicatorFields/indicatorfield-ACTI_Threat_Types.json
```
```json
{
  "id": "indicator_actithreattypes",
  "version": 1,
  "modified": "2022-03-07T10:26:49.471214293Z",
  "packID": "",
  "...":"...",
 }
 ```

- @ankitmordhwaj - UT for `ACTIThreatIntelReport.py` is still below 70% and failing. See [the coverage report](https://490377-60525392-gh.circle-artifacts.com/0/artifacts/coverage_report/html/d_cea6e15193f44573_ACTIThreatIntelReport_py.html).

File: Packs/AccentureCTI/IndicatorTypes/reputation-ACTI_Intelligence_Report.json [hunk 1]
```diff
@@ -0,0 +1,94 @@
+[
```
Comment (kgal-pan) [inline Packs/AccentureCTI/IndicatorTypes/reputation-ACTI_Intelligence_Report.json:1]: the JSON is supposed to be an `Object`, not `Array`. Remove square brackets.

File: Packs/AccentureCTI/IndicatorTypes/reputation-ACTI_Intelligence_Alert.json [hunk 2]
```diff
@@ -0,0 +1,106 @@
+[
```
Comment (kgal-pan) [inline Packs/AccentureCTI/IndicatorTypes/reputation-ACTI_Intelligence_Alert.json:1]: the JSON is supposed to be an `Object`, not `Array`. Remove square brackets.

**Comment by kgal-pan** [comment]
@satyakidroid - Please run the following commands to fix formatting of indicator types:

```bash
demisto-sdk format -i Packs/AccentureCTI/IndicatorTypes/reputation-ACTI_Intelligence_Report.json
demisto-sdk format -i Packs/AccentureCTI/IndicatorTypes/reputation-ACTI_Intelligence_Alert.json
```

Still waiting to increase coverage of `ACTIThreatIntelReport.py`.

**Comment by satyakidroid** [comment]
> @satyakidroid - Please run the following commands to fix formatting of indicator types:
> 
> ```shell
> demisto-sdk format -i Packs/AccentureCTI/IndicatorTypes/reputation-ACTI_Intelligence_Report.json
> demisto-sdk format -i Packs/AccentureCTI/IndicatorTypes/reputation-ACTI_Intelligence_Alert.json
> ```
> 
> Still waiting to increase coverage of `ACTIThreatIntelReport.py`.
I tried this, but was failing for some reason

**Comment by tomer-pan** [comment]
Hi @satyakidroid,
It seems that you have created some custom indicator fields which are identical (or very similar) to the common fields, such as:
Threat Types (same name), Last Modified (Modified), Last Published (Published), Report Title (Title), Severity (Source Original Severity). Please use the common indicator fields instead wherever possible since it is not ideal to create duplicated fields. Also, please add the prefix `acti`/`ACTI` to the custom fields' names, IDs and clinames.

**Comment by satyakidroid** [comment]
> Hi @satyakidroid, It seems that you have created some custom indicator fields which are identical (or very similar) to the common fields, such as: Threat Types (same name), Last Modified (Modified), Last Published (Published), Report Title (Title), Severity (Source Original Severity). Please use the common indicator fields instead wherever possible since it is not ideal to create duplicated fields. Also, please add the prefix `acti`/`ACTI` to the custom fields' names, IDs and clinames.

You're correct. But I was'nt able to tag to those fields as these predefined fields wasn't there in `Custom Fields` section in UI. I could only reuse `Creation Date` which was available.
image

**Comment by tomer-pan** [comment]
> > Hi @satyakidroid, It seems that you have created some custom indicator fields which are identical (or very similar) to the common fields, such as: Threat Types (same name), Last Modified (Modified), Last Published (Published), Report Title (Title), Severity (Source Original Severity). Please use the common indicator fields instead wherever possible since it is not ideal to create duplicated fields. Also, please add the prefix `acti`/`ACTI` to the custom fields' names, IDs and clinames.
> 
> You're correct. But I was'nt able to tag to those fields as these predefined fields wasn't there in `Custom Fields` section in UI. I could only reuse `Creation Date` which was available. image

You would need to associate the relevant fields to the new indicator types (as you did with the new fields) and then you will be able to map them as well.

**Comment by satyakidroid** [comment]
> > > Hi @satyakidroid, It seems that you have created some custom indicator fields which are identical (or very similar) to the common fields, such as: Threat Types (same name), Last Modified (Modified), Last Published (Published), Report Title (Title), Severity (Source Original Severity). Please use the common indicator fields instead wherever possible since it is not ideal to create duplicated fields. Also, please add the prefix `acti`/`ACTI` to the custom fields' names, IDs and clinames.
> > 
> > 
> > You're correct. But I was'nt able to tag to those fields as these predefined fields wasn't there in `Custom Fields` section in UI. I could only reuse `Creation Date` which was available. image
> 
> You would need to associate the relevant fields to the new indicator types (as you did with the new fields) and then you will be able to map them as well.

Yes, previously I actually tried doing that, first I add those fields to layout, then I added those custom layouts to the custom indicator types and then when I tried to map I didn't find the predefined fields.

Like I remember trying this with `published` & `modified` but wasn't able to tag to those fields, as those didn't show up, (though I attached those to layouts which are assigned to the indicator type as well).

**Comment by tomer-pan** [comment]
@satyakidroid Here is how you associate an indicator field to an indicator type (under "Add to indicator types"):
image
You would need to choose your indicator type and only then you will be able to map it inside the indicator type.

**Comment by satyakidroid** [comment]
I added the indicator type as you directed:
image
and then I navigated to `Indicator Type` but couldn't see the `modified` field:
image

**Comment by tomer-pan** [comment]
@satyakidroid Specifically the "Modified" field is a system field, please use the "Updated Date" instead and associate your indicators types to it as well (in the field json file, under `associatedTypes` key). For example:
```
"associatedTypes": [
        "ACTI Intelligence Alert"
]
```
Please do not change the `associatedToAll` key in this (or any other common) field.
Also, another thing I noticed, please change all of the CUSTOM fields to `"associatedToAll": false` and add the relevant association to your indicator type (as mentioned in the above example).

**Comment by satyakidroid** [comment]
> @satyakidroid Specifically the "Modified" field is a system field, please use the "Updated Date" instead and associate your indicators types to it as well (in the field json file, under `associatedTypes` key). For example:
> 
> ```
> "associatedTypes": [
>         "ACTI Intelligence Alert"
> ]
> ```
> 
> Please do not change the `associatedToAll` key in this (or any other common) field. Also, another thing I noticed, please change all of the CUSTOM fields to `"associatedToAll": false` and add the relevant association to your indicator type (as mentioned in the above example).

Yep `Updated Date` is available unlike `Modified` / `Published`
image

**Review by yucohen — CHANGES_REQUESTED** [review]
@satyakidroid Thanks for your contribution! noce work
Please see my comments

File: Packs/AccentureCTI/Integrations/ACTIThreatIntelReport/ACTIThreatIntelReport.yml [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+      type: String
+    - contextPath: IAIR.attachment_links
+      description: Provides with the document links related to the Intelligence Alert. This field is specific to Intelligence Alert, for example, 'https://intelgraph.idefense.com/rest/files/download/...'
+      type: String
+    - contextPath: IAIR.severity
+      description: Provides severity rating. This field is specific to Intelligence Alert, for example, '4'
+      type: String
+    - contextPath: IAIR.key
+      description: The key of the IAIR, for example, 'c0ea8f2f-372c-44d1-ad81-efde4971110e'
+      type: String
```
[the remark this answers is not in the document]
Comment (satyakidroid) [reply Packs/AccentureCTI/Integrations/ACTIThreatIntelReport/ACTIThreatIntelReport.yml:97]: nope its not. Its actually a unique identification ID inside our `IntelGraph` platform.

**Comment by satyakidroid** [comment]
Hi @tomer-pan , let me know if any other changes are required.

**Comment by satyakidroid** [comment]
Another thing is I wanted to have automatic enrichment of our indicator types `ACTI Intelligence Report` & `ACTI Intelligence Alert` , but somehow its not working though I've added the command under `Reputation command` tab.
image

**Comment by yucohen** [comment]
@satyakidroid Great demo! following demo comments:
1. Change the url value in the command to get the uuid only, and not the entire url.
2. Use a transformer to change the value of the indicator to be the uuid instead of the whole url.
3. Remove the unnecessary data from the layout (dbot score...).
4. Verify what the `key` in the response stands for, and remove it if it isn't necessary.

@tomer-pan @kgal-pan @Ni-Knight anything to add?

**Comment by satyakidroid** [comment]
I will need help on formatting, I am not able to visualize how this can work, the goal for this is to have value field change to `uuid` instead of `url` but can this value field change if we use `formatting script`?

**Comment by kgal-pan** [comment]
> the goal for this is to have value field change to `uuid` instead of `url` but can this value field change if we use `formatting script`? 

Yes, that is precisely the goal of the Formatting Script, from [documentation](https://docs.paloaltonetworks.com/cortex/cortex-xsoar/6-2/cortex-xsoar-threat-intel-management-guide/manage-indicators/understand-indicators/indicator-types/indicator-type-profile.html) I referenced:

> Formatting Script | Modifies **how the indicator displays in Cortex XSOAR** .Formatting scripts must be tagged `indicator-format` in order to appear in the dropdown for the indicator type. The formatting script **has one argument, input, the indicator value**. The input argument should be an array, in order to accept multiple inputs and return an entry-result per input. The entry-result per input can be a json array to create multiple indicators. If the entry-result is an empty string, it will be ignored and no indicator will be created.

You've essentially written the logic for the extracting the UUID from the URL already in `getThreatReport_command` method in `ACTIThreatIntelReport.py:96`:

image

You just need to take the URL in as an argument to the formatter script and do the same thing you're doing in the code above and return the UUID.

**Comment by satyakidroid** [comment]
@kgal-pan  I think I need to create a custom script to do that ?

**Comment by kgal-pan** [comment]
> @kgal-pan I think I need to create a custom script to do that ?

Correct, and put it in the `Scripts` folder.

**Comment by satyakidroid** [comment]
Thanks, I was able to get `uuid` in place of `url`.

File: Packs/AccentureCTI/Scripts/FormatACTIURL.yml [hunk 4]
```diff
@@ -0,0 +1,37 @@
+commonfields:
```
Comment (kgal-pan) [inline Packs/AccentureCTI/Scripts/FormatACTIURL.yml:1]: Please use `demisto-sdk split` to split the YAML and Python files.

File: Packs/AccentureCTI/Scripts/FormatACTIURL/FormatACTIURL.yml [hunk 5]
```diff
@@ -1,7 +1,6 @@
 commonfields:
```
Comment (kgal-pan) [inline Packs/AccentureCTI/Scripts/FormatACTIURL/FormatACTIURL.yml:1]: `FormatACTIURL.py` is missing in `Packs/AccentureCTI/Scripts/FormatACTIURL`

Comment (satyakidroid) [reply Packs/AccentureCTI/Scripts/FormatACTIURL/FormatACTIURL.yml:1]: @kgal-pan I might be wrong, but I see `FormatACTIURL.py` under `Packs/AccentureCTI/Scripts/FormatACTIURL`

File: Packs/AccentureCTI/Integrations/ACTIIndicatorQuery/ACTIIndicatorQuery.py [hunk 6]
```diff
@@ -334,24 +370,37 @@ def uuid_command(client: Client, args: dict, reliability: DBotScoreReliability,
...
+                ia.append(alert.split("/")[-1])
+        if type(reports) is dict:
+            reports = list(analysis_info.get('Intelligence Reports', []).values())
+            for report in reports:
+                ir.append(report.split("/")[-1])
+
+        context = {
+            "intelligence_alerts": ia,
+            "intelligence_reports": ir
+        }
```
Comment (yucohen) [inline Packs/AccentureCTI/Integrations/ACTIIndicatorQuery/ACTIIndicatorQuery.py:388]: Seems like you are using this functionality quite often. Consider adding a method that implements it and call this method instead (not mandatory)

**Comment by yucohen** [comment]
@satyakidroid Hi
What about removing the new integration and adding the new command to the existing one instead?

**Comment by satyakidroid** [comment]
> @satyakidroid Hi What about removing the new integration and adding the new command to the existing one instead?

I am on that, will be ready by Wednesday.

**Comment by kgal-pan** [comment]
@satyakidroid @nirmalneupane - Please remove `.venv` file from Pack as it's [causing a failure running the secrets validations](https://app.circleci.com/pipelines/github/demisto/content/152009/workflows/14dca161-d17c-42e4-8dc7-2915945e2f34/jobs/495872?invite=true#step-105-7).

**Comment by satyakidroid** [comment]
@kgal-pan @yucohen we have finished making necessary changes.

**Comment by kgal-pan** [comment]
> The CircleCI build failed again. @satyakidroid take a look at the build details [here](https://circleci.com/gh/demisto/content/496181?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) - and try and fix the issues so that we can merge your proposed changes as soon as possible.
> 
> #### Failed Build Steps
> * Validate Files and Yaml

@satyakidroid - No need to intervene in this one, it's a problem with the SDK that was already fixed.

**Comment by satyakidroid** [comment]
> > The CircleCI build failed again. @satyakidroid take a look at the build details [here](https://circleci.com/gh/demisto/content/496181?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) - and try and fix the issues so that we can merge your proposed changes as soon as possible.
> > #### Failed Build Steps
> > 
> > * Validate Files and Yaml
> 
> @satyakidroid - No need to intervene in this one, it's a problem with the SDK that was already fixed.

Okay @kgal-pan

**Comment by yucohen** [comment]
@satyakidroid Please add a unit-test for the `url-to-uuid` method in the `FormatACTIURL.py` file (to theFormatACTIURL), add a file named `FormatACTIURL_test.py` and add your unit test there.

**Pull request merged**